overlord/servicestate/quota_control_test.go

```diff
@@ -0,0 +1,322 @@
+	st.Unlock()
+
+	// create a quota group with no snaps to be the parent
+	err := s.mgr.CreateQuota("foo", "", nil, quantity.SizeGiB)
+	c.Assert(err, IsNil)
+
+	// now we can create a sub-quota
+	err = s.mgr.CreateQuota("foo2", "foo", []string{"test-snap"}, quantity.SizeGiB)
+	c.Assert(err, IsNil)
+
+	// check that the quota groups were created in the state
+	st.Lock()
+	defer st.Unlock()
+	m, err := servicestate.AllQuotas(st)
+	c.Assert(err, IsNil)
+	c.Assert(m, HasLen, 2)
+	for name, grp := range m {
+		switch name {
+		case "foo":
+			c.Assert(grp.Snaps, HasLen, 0)
+			c.Assert(grp.SubGroups, DeepEquals, []string{"foo2"})
+			c.Assert(grp.ParentGroup, Equals, "")
+		case "foo2":
+			c.Assert(grp.Snaps, DeepEquals, []string{"test-snap"})
+			c.Assert(grp.SubGroups, HasLen, 0)
+			c.Assert(grp.ParentGroup, Equals, "foo")
+		default:
+			c.Errorf("unexpected group %q in state", name)
+		}
+	}
+}
+
+func (s *quotaControlSuite) TestRemoveQuota(c *C) {
+	r := s.mockSystemctlCalls(c, []expectedSystemctl{
+		{
+			// called for new slice unit written by CreateQuota after we create
+			// the snap in state
+			expArgs: []string{"daemon-reload"},
+		},
+		{
+			// called for the deleted slice unit from RemoveQuota
+			expArgs: []string{"daemon-reload"},
+		},
+		{
+			// called for the modified service unit files from EnsureSnapServices
+			// TODO: this call should go away?
+			expArgs: []string{"daemon-reload"},
+		},
+	})
+	defer r()
+
+	st := s.state
+	st.Lock()
+	// setup the snap so it exists
+	snapstate.Set(s.state, "test-snap", s.testSnapState)
+	snaptest.MockSnapCurrent(c, testYaml, s.testSnapSideInfo)
+	st.Unlock()
+
+	// trying to remove a group that does not exist fails
+	err := s.mgr.RemoveQuota("not-exists")
+	c.Assert(err, ErrorMatches, `cannot remove non-existent quota group "not-exists"`)
+
+	// create a quota
+	err = s.mgr.CreateQuota("foo", "", []string{"test-snap"}, quantity.SizeGiB)
+	c.Assert(err, IsNil)
+
+	// check that the quota groups was created in the state
+	st.Lock()
+	defer st.Unlock()
+	m, err := servicestate.AllQuotas(st)
+	c.Assert(err, IsNil)
+	c.Assert(m, HasLen, 1)
+	for name, grp := range m {
+		switch name {
+		case "foo":
+			c.Assert(grp.Snaps, DeepEquals, []string{"test-snap"})
+			c.Assert(grp.SubGroups, HasLen, 0)
+			c.Assert(grp.ParentGroup, Equals, "")
+		default:
+			c.Errorf("unexpected group %q in state", name)
+		}
+	}
+
+	// remove the quota from the state
+	st.Unlock()
+	defer st.Lock()
+	err = s.mgr.RemoveQuota("foo")
+	c.Assert(err, IsNil)
+
+	st.Lock()
+	defer st.Unlock()
+	m, err = servicestate.AllQuotas(st)
+	c.Assert(err, IsNil)
+	c.Assert(m, HasLen, 0)
+}
+
+func (s *quotaControlSuite) TestUpdateQuotaGroupNotExist(c *C) {
+	opts := servicestate.QuotaGroupUpdate{}
+	err := s.mgr.UpdateQuota("non-existing", opts)
+	c.Check(err, ErrorMatches, `group "non-existing" does not exist`)
+}
+
+func (s *quotaControlSuite) TestUpdateQuotaChangeMemLimit(c *C) {
```

Comment:
we should check what happens if we go over the limit? basically at least some error case checks per function

Reply:
added this test